pep-0590.rst

```diff
@@ -0,0 +1,256 @@
+Post-History: 
+
+Abstract
+========
+
+This PEP introduces a new calling convention [1]_ for use by CPython and other software and tools in the CPython ecosystem.
+The new calling convention is a formalisation and extension of a calling convention already used internally by CPython.
+
+Rationale
+=========
+
+The choice of a calling convention impacts the performance and flexibility of code on either side of the call.
+Often there is tension between performance and flexibility.
+
+The current ``tp_call`` [2]_ calling convention is sufficiently flexible to cover all cases, but its performance is poor.
+The poor performance is largely a result of having to create intermediate tuples, and possibly intermediate dicts, during the call. 
+This is mitigated in CPython by including special-case code to speed up calls to Python and builtin functions.
+Unfortunately this means that other callables such as classes and third party extension objects are called using the 
+slower, more general ``tp_call`` calling convention.
+
+This PEP proposes that the calling convention used internally for Python and builtin functions is generalized and published
+so that all calls can benefit from better performance.
+
+Improved Performance
+--------------------
+
+The current ``tp_call`` calling convention requires creation of a tuple and, if there are any named arguments, a dictionary for every call.
+This is expensive. The proposed calling convention removes the need to create almost all of these temporary objects.
+Another source of inefficiency in the ``tp_call`` convention is that it has one function pointer per-class, rather than per-object. This is inefficient for calls to classes as several intermediate objects need to be created. For a user defined class, ``UserClass``, at least one intermediate object is created for each call in the sequence ``type.__call__``, ``object.__new__``, ``UserClass.__init__``.
+
+The new proposed calling convention is not fully general, but covers the large majority of calls.
+It is designed to remove the overhead of temporary object creation and multiple indirections.
+
+Specification
+=============
+
+The function pointer type
+-------------------------
+
+Calls are made through a function pointer taking the following parameters:
+
+* ``PyObject *callable``: The called object
+* ``Py_ssize_t n``: The number of arguments plus an optional offset flag for the first argument in vector.
+* ``PyObject **args``: A vector of arguments
+* ``PyTupleObject *kwnames``: A tuple of the names of the named arguments.
+
+This is implemented by the function pointer type:
+``typedef PyObject *(*vectorcall)(PyObject *callable, Py_ssize_t n, PyObject** args, PyTupleObject *kwnames);``
+
+Changes to the ``PyTypeObject``
+-------------------------------
+
+The a new slot called ``tp_vectorcall_offset`` is added. It has the type ``uint32_t``.
+A new flag is added, ``Py_TPFLAGS_HAVE_VECTOR_CALL``, which is set for any new PyTypeObjects that include the
```

Comment:
`Py_TPFLAGS_HAVE_VECTOR_CALL` -> `Py_TPFLAGS_HAVE_VECTORCALL` to be consistent with `tp_vectorcall_offset` and other uses of "vectorcall" (not "vector call")